### PR TITLE
server: set correct gateway when update vm nic on shared networks

### PR DESCRIPTION
**Pull request by weizhouapache**
### Description

This PR fixes the issue that nic has wrong gateway after updating vm nic.

Steps to reproduce the issue
(1) create shared network (in advanced zone or advanced zone with sg)
(2) create new shared network (with same startip/endip/netmask, but different gateway).
(3) create a vm in new network
(4) stop vm and update vm nic ip address

Expected result:
The vm has correct gateway and netmask (of second network)

Actual result:
The vm has wrong gateway and netmask (of first network)

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [X] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
unit test failed, I will fix it.

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@blueorangutan test centos7 vmware-67u3

File: server/src/test/java/com/cloud/vm/UserVmManagerTest.java [hunk 1]
```diff
@@ -630,9 +631,9 @@ public void testUpdateVmNicIpSuccess2() throws Exception {
         when(vlan.getVlanNetmask()).thenReturn("255.255.255.0");
 
         when(_ipAddrMgr.allocatePublicIpForGuestNic(Mockito.eq(_networkMock), nullable(Long.class), Mockito.eq(_accountMock), anyString())).thenReturn("10.10.10.10");
-        lenient().when(_ipAddressDao.findByIpAndSourceNetworkId(anyLong(), anyString())).thenReturn(null);
+        when(_ipAddressDao.findByIpAndSourceNetworkId(anyLong(), eq("10.10.10.10"))).thenReturn(newIp);
+        when(_ipAddressDao.findByIpAndSourceNetworkId(anyLong(), eq("10.10.10.9"))).thenReturn(null);
```
Comment (borisstoyanov) [inline server/src/test/java/com/cloud/vm/UserVmManagerTest.java:635]: Sorry I hope I'm wrong here, but isn't this too specific, looks like hardcoded to me?

Comment (weizhouapache) [reply server/src/test/java/com/cloud/vm/UserVmManagerTest.java:635]: @borisstoyanov yes, it is hardcoded, but it is just an unit test.

**Comment by rhtyd** [comment]
cc @vladimirpetrov @shwstppr @Pearl1594

**Comment by rhtyd** [comment]
- Testing on basic zone w/without SG
- Testing on Adv zone w/without SG + shared

**Comment by weizhouapache** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@blueorangutan test centos7 vmware-67u3

**Comment by rhtyd** [comment]
@blueorangutan test

**Review by Pearl1594 — APPROVED** [review]
Code LGTM

**Review by shwstppr — APPROVED** [review]
Code LGTM

**Comment by andrijapanicsb** [comment]
> 
> 
>     * Testing on basic zone w/without SG
> 
>     * Testing on Adv zone w/without SG + shared

Can't have more than 1 network in Basic Zone - so my test effort is done - nothing to test :)

**Comment by weizhouapache** [comment]
> > ```
> > * Testing on basic zone w/without SG
> > 
> > * Testing on Adv zone w/without SG + shared
> > ```
> 
> Can't have more than 1 network in Basic Zone - so my test effort is done - nothing to test :)

@andrijapanicsb ah ok. Can you test if vm nic ip address can be changed when vm is stopped in basic zone ?

**Comment by andrijapanicsb** [comment]
yes @weizhouapache - just tested changing the IP on the stopped VM - all good, new IP assigned via DHCP, default route still as before that change - looks good.

I won't LGTM this one, as I haven't tested anything meaningful myself, but @vladimirpetrov did

**Review by vladimirpetrov — APPROVED** [review]
Tested with advanced zone with and without security groups - LGTM!

**Comment by weizhouapache** [comment]
thanks for review and testing ! @vladimirpetrov @andrijapanicsb @Pearl1594 @shwstppr

**Pull request merged**